### ISSUE TITLE
Extremely Sparse Datasets cannot be used for training

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#536]
This issue is identical to this one, but for LightGBM this time: https://github.com/dmlc/xgboost/issues/2326

The issue: LightGBM does not start training after a while (stuck in building matrix?)

Before running, to know:

* Dataset download is 240MB, uncompressed is 2GB, sum of RDS files are 650MB
* Creating the dataset (3GB, 7,744,201,107,060 elements) requires using my svmlight package in R
* My scripts can run copy&paste just by changing the folders appropriately in `setwd`

Created files info:

```
  File: url_svmlight.tar.gz
CRC-32: c152f632
   MD4: b05d0a58ad6f53f9ad20a5f350c25df2
   MD5: 2eb74f59d06807a3845f90ed36fe8ded
 SHA-1: 605121dde12ce7baf098e30cd8856ef4ed6d5e69
```

```
  File: reput_sparse.rds
CRC-32: 846b1907
   MD4: 2ea36b208118b7d46d18ccb5fad8da98
   MD5: ce371f418463ebc799367ed41f5fd654
 SHA-1: 7f2a73609d72ff61a24a6d176c44752b68954ea6
```

## Environment info
Operating System: Windows Server 2012 R2 (baremetal) - OS doesn't matter, happens also in Ubuntu 17.04

Compiler: MinGW 7.1

Package used (python/R/jvm/C++): R 3.4

`LightGBM` version used: from source, latest commit of today (e5e7217)

## Computer specs:
* Quad E7-88xx v4 (72 cores), but 3 CPUs were disabled to get rid of NUMA, and I kept only 4 cores online (14 cores disabled)
* 1TB RAM 1866MHz RAM (256GB RAM available to get rid of NUMA)
* Not virtualized

Used 8 threads for LightGBM because it was for a benchmark comparison.

## Steps to reproduce

1. Download this dataset: http://archive.ics.uci.edu/ml/datasets/URL+Reputation (svmlight format, 121 parts, 2396130 rows, 3231961 features
2. Load data using my svmlight parser
3. Train a LightGBM model (default is 255 bins, one can start directly with 15 bins I think)

Scripts:

LIBRARY DOWNLOAD:

```r
install.packages("devtools")
install.packages("Matrix")
install.packages("Rcpp")
install.packages("RcppEigen")
devtools::install_github("Laurae2/sparsity")
# Install LightGBM separately from source
```

DATA LOAD:

```r
# Libraries
library(sparsity)
library(Matrix)

# SET YOUR WORKING DIRECTORY
setwd("E:/benchmark_lot/data")

data <- read.svmlight(paste0("Day0.svm"))
data$matrix@Dim[2] <- 3231962L
data$matrix@p[length(data$matrix@p):3231963] <- data$matrix@p[length(data$matrix@p)]
data$matrix <- data$matrix[1:(data$matrix@Dim[1] - 1), ]
label <- (data$labels[1:(data$matrix@Dim[1])] + 1) / 2
data <- data$matrix

new_data <- list()

for (i in 1:120) {
  indexed <- (i %% 10) + (10 * ((i %% 10) == 0))
  new_data[[indexed]] <- read.svmlight(paste0("Day", i, ".svm"))
  new_data[[indexed]]$matrix@Dim[2] <- 3231962L
  new_data[[indexed]]$matrix@p[length(new_data[[indexed]]$matrix@p):3231963] <- new_data[[indexed]]$matrix@p[length(new_data[[indexed]]$matrix@p)]
  new_data[[indexed]]$matrix <- new_data[[indexed]]$matrix[1:(new_data[[indexed]]$matrix@Dim[1] - 1), ]
  label <- c(label, (new_data[[indexed]]$labels[1:(new_data[[indexed]]$matrix@Dim[1])] + 1) / 2)
  
  if ((i %% 10) == 0) {
    
    data <- rbind(data, new_data[[1]]$matrix, new_data[[2]]$matrix, new_data[[3]]$matrix, new_data[[4]]$matrix, new_data[[5]]$matrix, new_data[[6]]$matrix, new_data[[7]]$matrix, new_data[[8]]$matrix, new_data[[9]]$matrix, new_data[[10]]$matrix)
    gc(verbose = FALSE)
    
    cat("Parsed element 'Day", i, ".svm'. Sparsity: ", sprintf("%05.0f", as.numeric(data@Dim[1]) * as.numeric(data@Dim[2]) / length(data@i)), ":1. Balance: ", sprintf("%04.02f", length(label) / sum(label)), ":1.\n", sep = "")
    
  }
  
}

# Save to RDS
gc()
saveRDS(data, file = "reput_sparse.rds", compress = TRUE)

# Save labels
saveRDS(label, file = "reput_label.rds", compress = TRUE)
```

RUN LightGBM, will take forever(?) to start:

```r
# SET YOUR WORKING DIRECTORY
setwd("E:/benchmark_lot/data")

my_data <- readRDS("reput_sparse.rds")
label <- readRDS("reput_label.rds")

library(lightgbm)
my_train <- my_data[1:2250000, ]
my_train_label <- label[1:2250000]
my_test <- my_data[2250001:2396130, ]
my_test_label <- label[2250001:2396130]

train <- lgb.Dataset(data = my_train, label = my_train_label)
test <- lgb.Dataset(data = my_test, label = my_test_label)

gc()
model <- lgb.train(params = list(objective = "binary",
                                 metric = "auc",
                                 bin_construct_sample_cnt = 2250000L,
                                 early_stopping_round = 25),
                   train,
                   10000,
                   list(test = test),
                   verbose = 2)
```

**Comment by guolinke** [comment]
@Laurae2  reduce ```bin_construct_sample_cnt''' ?

**Comment by guolinke** [comment]
@Laurae2 I had tested on  "kdd2010 (bridge to algebra)" dataset (https://www.csie.ntu.edu.tw/~cjlin/libsvmtools/datasets/binary.html) . 
its # of data: 19,264,097 and # of features: 29,890,095.  The data loading finished in 200 seconds (bin_construct_sample_cnt=600k, min_data=1000) .

**Comment by guolinke** [comment]
@Laurae2 any updates ?

**Comment by Laurae2** [comment]
@guolinke I used this but it still didn't start training after 10 minutes:

* Number of observations: 1,000,000 (approx 28000:1 sparsity)
* Number of features: 3,231,961 (left untouched)
* `bin_construct_sample_cnt = 100000`
* `min_data = 1000`

It was tested using my laptop (dual core i7-4600U, 2.7GHz) as I don't have access to my server currently. All cores busy 100%. Maybe I missed something in the parameters? This is my script:

```r
setwd("D:/Data Science/url_grading/url_svmlight")

my_data <- readRDS("../reput_sparse.rds")
label <- readRDS("../reput_label.rds")

library(lightgbm)
my_train <- my_data[1:1000000, ]
my_train_label <- label[1:1000000]
my_test <- my_data[1000001:1196130, ]
my_test_label <- label[1000001:1196130]
gc()

train <- lgb.Dataset(data = my_train, label = my_train_label)
test <- lgb.Dataset(data = my_test, label = my_test_label)

gc()
model <- lgb.train(params = list(objective = "binary",
                                 metric = "auc",
                                 bin_construct_sample_cnt = 100000L,
                                 min_data = 1000,
                                 early_stopping_round = 5),
                   train,
                   10,
                   list(test = test),
                   verbose = 2)
```

**Comment by Laurae2** [comment]
@guolinke It eventually started training after 15 minutes and took less than 10 seconds for the 10 iterations. But it used only 2,807 features when I want all 3,231,961 features:

```r
[LightGBM] [Info] Number of positive: 334925, number of negative: 665075
[LightGBM] [Info] Total Bins 7726
[LightGBM] [Info] Number of data: 1000000, number of used features: 2807
```

I think I'll parse the matrix myself before feeding it to LightGBM so I know what is inside and what is not inside.

Number of elements in features by index:

![image](https://cloud.githubusercontent.com/assets/9083669/26282661/f88b6124-3e15-11e7-8cd9-834827625acb.png)

By the way, is it possible to do the building part (which happens before the training) with a specified number of threads? When testing 72 parallel trainings (each with `num_threads = 1`) on the machine with 144 threads, LightGBM spawned 10000+ threads resulting in a kernel hard lockup (eventually crashing).

**Comment by guolinke** [comment]
@Laurae2 
Actually, it is not need to use all features, The features will be ignored in following condition:
1. Not show in the sampling data for the bin construction. LightGBM uses sampled data to construct bin mapper. If a feature doesn't show up in sampling data, it means this feature is nearly all zeros, ignoring it won't have much effect for the accuracy. 
2. the number of nonzero values smaller than ```min_data```. It means this feature cannot be split. 

However, if you still need to use more features, you can set a larger ```bin_construct_sample_cnt``` and smaller ```num_data```. But it won't have much effect on the accuracy. 

For the number of threads setting, I just check the code. It didn't enable for data loading. I will fix it.

**Comment by Laurae2** [comment]
@guolinke CPU usage is correct with 1 thread (out of 4 threads) with your commit 0b2b213:

![image](https://cloud.githubusercontent.com/assets/9083669/26282908/5a7c75e4-3e1b-11e7-82b2-78c2da538e8b.png)

![image](https://cloud.githubusercontent.com/assets/9083669/26282907/52a8b4ae-3e1b-11e7-8a0d-77a28e711abc.png)
